### PR TITLE
Mesh shader refactoring and bugfixes

### PR DESCRIPTION
**Pull request by expipiplus1**
Closes https://github.com/shader-slang/slang/issues/2804 and https://github.com/shader-slang/slang/issues/3401

### CODE REVIEW

File: source/slang/core.meta.slang [hunk 1]
```diff
@@ -1255,37 +1255,44 @@ struct ParameterBlock {}
...
 {
     __subscript(uint index) -> T
     {
-        // TODO: Ellie: It's illegal to not write out the whole primitive at once, should we use set over ref?
-        __intrinsic_op($(kIROp_GetElementPtr))
-        ref;
+        // It's illegal to not write out the entire primitive at once, so limit
+        // this to set
+        [mutating]
+        __intrinsic_op($(kIROp_MeshOutputSet))
```
Comment (csyonghe) [inline source/slang/core.meta.slang:1282]: We maybe able to get rid of this opcode by having `set` call `__getMeshOutputRef() = newValue;`

File: source/slang/slang-ir-glsl-legalize.cpp [hunk 2]
```diff
@@ -2094,106 +2094,118 @@ static void legalizeMeshOutputParam(
...
+                //   vertices[n] = tmp;
+                //
+                // This has copy-out semantics, which is really the
+                // best we can hope for without going and
+                // specializing foo.
+                auto ptr = as<IRPtrTypeBase>(a->getFullType());
+                SLANG_ASSERT(ptr && "Mesh output parameter was passed by value");
+                auto t = ptr->getValueType();
+                auto tmp = builder->emitVar(t);
+                for(UInt i = 0; i < c->getOperandCount(); i++)
```
Comment (csyonghe) [inline source/slang/slang-ir-glsl-legalize.cpp:2184]: Can we use `c->getArg` and `c->setArg` instead?

File: tests/pipeline/rasterization/mesh/hlsl-syntax.slang [hunk 3]
```diff
@@ -26,7 +26,7 @@ const static uint MAX_VERTS = 3;
 const static uint MAX_PRIMS = 1;
 
 // Test that we can convert the HLSL Syntax to the typed syntax
-void foo(uint tig, out Vertices<Vertex, MAX_VERTS> verts)
+void foo(uint tig, OutputVertices<Vertex, MAX_VERTS> verts)
```
Comment (csyonghe) [inline tests/pipeline/rasterization/mesh/hlsl-syntax.slang:29]: Why are we removing the test for the `out Vertices` syntax?

**Review by csyonghe — APPROVED** [review]

**Comment by Ipotrick** [comment]
This looks close to beeing mergable. I have great interest in using the enhancements in this pr.

**Pull request merged**